src/types/boxed.rs

```diff
@@ -0,0 +1,343 @@
+use std::any::{self, Any};
+use std::ops::Deref;
+
+use neon_runtime::raw;
+use neon_runtime::external;
+
+use crate::context::{Context, FinalizeContext};
+use crate::context::internal::Env;
+use crate::handle::{Managed, Handle};
+use crate::types::internal::ValueInternal;
+use crate::types::Value;
+
+type BoxAny = Box<dyn Any + Send + 'static>;
+
+/// A smart pointer for Rust data managed by the JavaScript engine.
+///
+/// The type `JsBox<T>` provides shared ownership of a value of type `T`,
+/// allocated in the heap. The data is owned by the JavaScript engine and the
+/// lifetime is managed by the JavaScript garbage collector.
+///
+/// Shared references in Rust disallow mutation by default, and `JsBox` is no
+/// exception: you cannot generally obtain a mutable reference to something
+/// inside a `JsBox`. If you need to mutate through a `JsBox`, use
+/// [`Cell`](https://doc.rust-lang.org/std/cell/struct.Cell.html),
+/// [`RefCell`](https://doc.rust-lang.org/stable/std/cell/struct.RefCell.html),
+/// or one of the other types that provide interior mutability.
+///
+/// ## `Deref` behavior
+///
+/// `JsBox<T>` automatically dereferences to `T` (via the `Deref` trait), so
+/// you can call `T`'s method on a value of type `JsBox<T>`.
+///
+/// ```rust
+/// # use neon::prelude::*;
+/// # fn my_neon_function(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+/// let vec: Handle<JsBox<Vec<_>>> = cx.boxed(vec![1, 2, 3]);
+///
+/// println!("Length: {}", vec.len());
+/// # Ok(cx.undefined())
+/// # }
+/// ```
+///
+/// ## Examples
+///
+/// Passing some immutable data between Rust and JavaScript.
+///
+/// ```rust
+/// # use neon::prelude::*;
+/// # use std::path::{Path, PathBuf};
+/// fn create_path(mut cx: FunctionContext) -> JsResult<JsBox<PathBuf>> {
+///     let path = cx.argument::<JsString>(0)?.value(&mut cx);
+///     let path = Path::new(&path).to_path_buf();
+///
+///     Ok(cx.boxed(path))
+/// }
+///
+/// fn print_path(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+///     let path = cx.argument::<JsBox<PathBuf>>(0)?;
+///
+///     println!("{}", path.display());
+///
+///     Ok(cx.undefined())
+/// }
+/// ```
+///
+/// Passing a user defined struct wrapped in a `RefCell` for mutability. This
+/// pattern is useful for creating classes in JavaScript.
+///
+/// ```rust
+/// # use neon::prelude::*;
+/// # use std::cell::RefCell;
+///
+/// type BoxedPerson = JsBox<RefCell<Person>>;
+///
+/// struct Person {
+///      name: String,
+/// }
+///
+/// impl Finalize for Person {}
```

Comment:
I think we haven't documented the `Finalize` trait and how it's connected to `JsBox` anywhere in this PR?